src/sources/soundsourcestem.cpp

```diff
@@ -0,0 +1,468 @@
+#include "sources/soundsourcestem.h"
+
+#include "sources/readaheadframebuffer.h"
+
+extern "C" {
+
+#include <libavutil/avutil.h>
+#if LIBAVUTIL_VERSION_INT >= AV_VERSION_INT(57, 28, 100) // FFmpeg 5.1
+#include <libavutil/channel_layout.h>
+#endif
+
+} // extern "C"
+
+#include "util/assert.h"
+#include "util/logger.h"
+#include "util/sample.h"
+
+#if !defined(VERBOSE_DEBUG_LOG)
+#define VERBOSE_DEBUG_LOG false
+#endif
+
+namespace mixxx {
+
+namespace {
+
+// STEM constants
+constexpr int kNumStreams = 5;
+constexpr int kRequiredStreamCount = kNumStreams - 1; // Stem count doesn't include the main mix
+
+const Logger kLogger("SoundSourceSTEM");
+
+} // anonymous namespace
+
+const QString SoundSourceProviderSTEM::kDisplayName = QStringLiteral("STEM with FFmpeg");
+
+QStringList SoundSourceProviderSTEM::getSupportedFileTypes() const {
+    return {"stem.mp4", "stem.m4a"};
+}
+
+SoundSourceProviderPriority SoundSourceProviderSTEM::getPriorityHint(
+        const QString& supportedFileType) const {
+    Q_UNUSED(supportedFileType)
+    return SoundSourceProviderPriority::Higher;
+}
+
+QString SoundSourceProviderSTEM::getVersionString() const {
+    return QString::fromUtf8(av_version_info());
+}
+
+SoundSourceSingleSTEM::SoundSourceSingleSTEM(const QUrl& url, unsigned int streamIdx)
+        : SoundSourceFFmpeg(url), m_streamIdx(streamIdx) {
+}
+
+SoundSource::OpenResult SoundSourceSingleSTEM::tryOpen(
```

Comment:
Is there an early check if this is really a stem file? If not, it should fail in the first attempt. Maybe also later and the normal m4a decoder can take over.

Reply:
There is a check that ensure the audio tracks have matching spec. Is there other checks you would like to see?

Reply:
I have no specific demand. Just keep in mind that this tryOpen() is called according to the priority of a certain format.
In other sound sources there is a Strict and a Permissive mode. I have no idea if this here reasonable as well.

Reply:
wouldn't it be possible to try to create a stem soundsource for any mp4, and if we don't find the metadata, or we face other validation error, continue with the generic mp4 reader?

Reply:
We could do that but that feels wrong and biased to assume that MP4 == STEM IMO. What is the the team opinion?

Reply:
We have a probing facility already. 
If the file is a stem.mp4 it is probed here first. 
In case tryOpen() fails, libfaad should take over.